Add command "exec_progbuf"

JanMatCodasip opened the pull request:
Command "exec_progbuf" allows to execute a user-specified sequence of instructions using the program buffer.

Change-Id: If3b9614129d0b6fcbc33fade29d3d60b35e52f98
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Add command "exec_progbuf"

Command "exec_progbuf" allows to execute a user-specified sequence
of instructions using the program buffer.

Change-Id: If3b9614129d0b6fcbc33fade29d3d60b35e52f98
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip commented:
This `exec_progbuf` command is intended to allow users to execute their own instructions using program buffer - without the need to craft these opreations using low-level commands `riscv dmi_read` and `riscv dmi_write`.

The idea to have this utility command had been mentioned some time ago here: https://github.com/riscv/riscv-openocd/pull/678#pullrequestreview-882005735

timsifive commented on doc/openocd.texi:
```diff
@@ -10753,6 +10753,19 @@ For details on this trigger type, see the RISC-V Debug Specification.
+@deffn {Command} {riscv exec_progbuf inst1 [inst2 [... inst_n]]}
+Execute the given sequence of instructions on the target using the Program Buffer.
+The instructions @var{inst1} .. @var{inst_n} shall be specified in their binary form
+(as 32-bit intergers). The terminating @var{ebreak} instruction needs not be specified -
```
```suggestion
(as 32-bit integers). The terminating @var{ebreak} instruction needs not be specified -
```

JanMatCodasip replied:
Thank you - fixed.

timsifive commented on doc/openocd.texi:
```diff
@@ -10753,6 +10753,19 @@ For details on this trigger type, see the RISC-V Debug Specification.
+can use to execute sequences of arbitrary instructions (small programs) on the target.
+For details on the Program Buffer, please refer to the RISC-V Debug Specification.
+
+@deffn {Command} {riscv exec_progbuf inst1 [inst2 [... inst_n]]}
```
Is it worth describing how compressed instructions fit into the 32-bit words?

JanMatCodasip replied:
I believe that would be useful - I have added that information along with examples. I have made minor reword of the documentation, too.

timsifive reviewed (dismissed):
Looks good.

For my personal curiosity, what are you using this command for?

JanMatCodasip committed:
Updated the doc:

- Minor reword and reorder of the sentences.
- Added information about C-instructions in progbuf.
- Fixed a typo (per the review).
- Added examples.

Change-Id: I88c9a3ff3c6b60614be7eafd3a6f21be722a77b7
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip commented:
Thank you for the review.

> what are you using this command for?

We use that command mainly for testing and for execution of fences in some cases. But it is a generic command - I believe  people will find various other uses for it.

JanMatCodasip committed:
Cosmetic changes

Change-Id: I7135c9f435f640e189c7d7922a2702814dfd595f
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

timsifive reviewed (approved):
Looks good!
I agree that this seems generally useful. I was just curious what specific use you have. Testing and fences makes sense to me.
